src/filesystem/web.rs

```diff
@@ -0,0 +1,1082 @@
+    pub async fn from_idb_key(idb_key: String) -> Option<Self> {
+        if !Self::filesystem_supported() {
+            return None;
+        }
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        filesystem_tx_or_die()
+            .send(FileSystemCommand(FileSystemCommandInner::DirFromIdb(
+                idb_key.clone(),
+                oneshot_tx,
+            )))
+            .unwrap();
+        oneshot_rx.await.unwrap().map(|(key, name)| FileSystem {
+            key,
+            name,
+            idb_key: Some(idb_key),
+        })
+    }
+
+    /// Creates a new `FileSystem` from a subdirectory of this one.
+    pub fn subdir(&self, path: impl AsRef<camino::Utf8Path>) -> Result<Self, Error> {
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        filesystem_tx_or_die()
+            .send(FileSystemCommand(FileSystemCommandInner::DirSubdir(
+                self.key,
+                path.as_ref().to_path_buf(),
+                oneshot_tx,
+            )))
+            .unwrap();
+        oneshot_rx
+            .blocking_recv()
+            .unwrap()
+            .map(|(key, name, idb_key)| FileSystem { key, name, idb_key })
+    }
+
+    /// Drops the directory with the given key from IndexedDB if it exists in there.
+    pub fn idb_drop(idb_key: String) -> bool {
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        filesystem_tx_or_die()
+            .send(FileSystemCommand(FileSystemCommandInner::DirIdbDrop(
+                idb_key, oneshot_tx,
+            )))
+            .unwrap();
+        oneshot_rx.blocking_recv().unwrap()
+    }
+
+    /// Returns a path consisting of a single element: the name of the root directory of this
+    /// filesystem.
+    pub fn root_path(&self) -> &camino::Utf8Path {
+        self.name.as_str().into()
+    }
+
+    /// Returns the key needed to restore this `FileSystem` using `FileSystem::from_idb()`.
+    pub fn idb_key(&self) -> Option<&str> {
+        self.idb_key.as_deref()
+    }
+}
+
+impl Drop for FileSystem {
+    fn drop(&mut self) {
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        filesystem_tx_or_die()
+            .send(FileSystemCommand(FileSystemCommandInner::DirDrop(
+                self.key, oneshot_tx,
+            )))
+            .unwrap();
+        oneshot_rx.blocking_recv().unwrap();
+    }
+}
+
+impl Clone for FileSystem {
+    fn clone(&self) -> Self {
+        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        filesystem_tx_or_die()
+            .send(FileSystemCommand(FileSystemCommandInner::DirClone(
+                self.key, oneshot_tx,
+            )))
+            .unwrap();
+        Self {
+            key: oneshot_rx.blocking_recv().unwrap(),
+            name: self.name.clone(),
+            idb_key: self.idb_key.clone(),
+        }
+    }
+}
+
+impl FileSystemTrait for FileSystem {
+    type File<'fs> = File where Self: 'fs;
+
+    fn open_file(
```

Comment:
Saving a project has the error "permission denied"
![image](https://github.com/Astrabit-ST/Luminol/assets/54482069/3e9c7079-652d-435c-9d9f-b4b16803f2f2)